doc/ir-semantics.text

```diff
@@ -111,18 +111,15 @@ Update (partial) function, returns an H:
 
 ## Osr transformation of environment
 
-  materialize P H E (mut x = e) = (x ↦ (a ↦ v))
-    where a - fresh
-      and v := eval P H E e
-  materialize P H E _           = ()
-
-  osr-env P H E M (const x = e) = (x ↦ val v)
+  create-env H E (const x = e) = (∅, [x ↦ val v])
     where v := eval P H E e
-  osr-env P H E M (mut x)       = (x ↦ ⊥)
-  osr-env P H E M (mut x = &y)  = (x ↦ ref a)
+  create-env H E (mut x)       = ([a ↦ ⊥], [x ↦ ref a])
+    where a fresh
```

Comment:
So far we stored ⊥ in the env. See the definition of envs `val v | ref a | ⊥`. If you want to store it in the heap then we'd have to update the definition of `E` and potentially other places too. I lean towards keeping it in the env. The reason is that (1) it's not necessary to create a heap cell. And (2) it should never be allowed to read an undefined variable. For example with your changes the following program would not error on line 2 but I think it should.

```
  mut x
  call foo(&x)

function foo(mut x)
  return 0
```